### ISSUE TITLE
Passing `src_key_padding_mask` as `bool` vs `float` causes different outputs from `nn.TransformerEncoderLayer`

### ISSUE THREAD

**Issue by petercall** [pytorch/pytorch#147209]
### 🐛 Describe the bug

The module `nn.TransformerEncoderLayer` is outputting different values from the `forward` function based on the data type of the `src_key_padding_mask` argument. The documentation says that either boolean data type or float data type should be accepted, so the difference in output is puzzling.

```python
import torch
import torch.nn as nn

batch_size = 1
context_length = 4
embed_dim = 4
nheads=1

model = nn.TransformerEncoderLayer(embed_dim, nheads, dropout = 0, batch_first = True, norm_first = True)

input_tensor = torch.randn(batch_size, context_length, embed_dim)
padding_mask = torch.randint(low=0, high = 2, size = (batch_size, context_length))


bool_output = model(input_tensor, src_key_padding_mask = padding_mask.to(torch.float32))
float_output = model(input_tensor, src_key_padding_mask = padding_mask.to(torch.bool))

print(torch.allclose(bool_output, float_output, equal_nan = True, atol=1e-05))

```
```
output: False
```

I coded up a manual version of `nn.TransformerEncoderLayer` by hand, and mine is matching the output when the input is of dtype `bool`. However, I also noted that when I also include causal masking, the `float` implementation is able to handle when an entire row in the attention matrix equals `-inf`, and the `bool` implementation (as well as my manual implementation) returns `nans` for those rows. I believe that the `float` implementation has had some fix applied to handle when an entire row is set to `-inf`, but the `bool` implementation has not. I think the `bool` implementation needs to be updated to also handle this case. As to why they are outputting different values, I think it could be related to the fix applied to the `float` implementation. Ultimately we don't want the model to return rows of `nans`, so I favor the `float` implementation, but it should be fixed so that the output of `nn.TransformerEconderLayer` is the same regardless of the dtype of `src_key_padding_mask`.

### Versions

Collecting environment information...
PyTorch version: 2.4.0
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.12.4 | packaged by Anaconda, Inc. | (main, Jun 18 2024, 15:12:24) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-52-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.64
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA RTX A5000
Nvidia driver version: 550.120
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               36
On-line CPU(s) list:                  0-35
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) W-2295 CPU @ 3.00GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   2
Core(s) per socket:                   18
Socket(s):                            1
Stepping:                             7
CPU max MHz:                          4800.0000
CPU min MHz:                          1200.0000
BogoMIPS:                             6000.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req vnmi avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            576 KiB (18 instances)
L1i cache:                            576 KiB (18 instances)
L2 cache:                             18 MiB (18 instances)
L3 cache:                             24.8 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-35
Vulnerability Gather data sampling:   Mitigation; Microcode
Vulnerability Itlb multihit:          KVM: Mitigation: VMX disabled
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.4.0
[pip3] torchaudio==2.4.0
[pip3] torchvision==0.19.0
[pip3] triton==3.0.0
[conda] blas                      1.0                         mkl  
[conda] cuda-cudart               12.4.127                      0    nvidia
[conda] cuda-cupti                12.4.127                      0    nvidia
[conda] cuda-libraries            12.4.0                        0    nvidia
[conda] cuda-nvrtc                12.4.127                      0    nvidia
[conda] cuda-nvtx                 12.4.127                      0    nvidia
[conda] cuda-opencl               12.6.37                       0    nvidia
[conda] cuda-runtime              12.4.0                        0    nvidia
[conda] cudatoolkit-dev           11.7.0               h1de0b5d_6    conda-forge
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libcublas                 12.4.2.65                     0    nvidia
[conda] libcufft                  11.2.0.44                     0    nvidia
[conda] libcurand                 10.3.7.37                     0    nvidia
[conda] libcusolver               11.6.0.99                     0    nvidia
[conda] libcusparse               12.3.0.142                    0    nvidia
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch
[conda] libnvjitlink              12.4.99                       0    nvidia
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py312h5eee18b_1  
[conda] mkl_fft                   1.3.8           py312h5eee18b_0  
[conda] mkl_random                1.2.4           py312hdb19cb5_0  
[conda] numpy                     1.26.4          py312hc5e2394_0  
[conda] numpy-base                1.26.4          py312h0da6c21_0  
[conda] pytorch                   2.4.0           py3.12_cuda12.4_cudnn9.1.0_0    pytorch
[conda] pytorch-cuda              12.4                 hc786d27_6    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.4.0               py312_cu124    pytorch
[conda] torchtriton               3.0.0                     py312    pytorch
[conda] torchvision               0.19.0              py312_cu124    pytorch

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by ngimel** [comment]
This is expected, see https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html#torch.nn.MultiheadAttention.forward, docs for `key_padding_mask`

**Comment by petercall** [comment]
Thank you for the clarification!

**Comment by colesbury** [comment]
Closing as this sounds like it's the expected behavior.
